File: LeetCode/practice-cpp/LeetCode-5-Manacher.cpp
// 最长回文子串
class Solution {
public:
    string longestPalindrome(string s) {
        if (s.size() <= 1) return s;

        string str(2 * s.size() + 1, '#');
        for (int i = 0; i < s.size(); ++i) {
            str[2 * i + 1] = s[i];
        }
        
        vector<int> armLens(str.size(), 0);
        int bestCenter = 0, bestArmLen = -1; // 记录最大回文窗口
        int currCenter = 0, currArmLen = -1; // 记录当前回文窗口

        for (int i = 0; i < str.size(); ++i) {
            int armLen = 0;
            if (i > currCenter + currArmLen) { // 索引在臂长边界外，直接中心探索
                // armLen = 0;
            } else if (i + armLens[2 * currCenter - i] <= currCenter + currArmLen) { // 索引在臂长边界内，加上对称位置臂长也在臂长边界内
                armLen = armLens[2 * currCenter - i];
            } else { // 索引在臂长边界内，加上对称位置臂长在臂长边界外
                armLen = currCenter + currArmLen - i;
            }

            while (i - armLen - 1 >= 0 && i + armLen + 1 < str.size() &&
                   str[i - armLen - 1] == str[i + armLen + 1]) {
                ++ armLen;
            }
            armLens[i] = armLen;

            if (i + armLen > currCenter + currArmLen) { // 更新臂长可达的最大索引和中心
                currCenter = i;
                currArmLen = armLen;
            }
            if (armLen > bestArmLen) { // 更新最大臂长和中心
                bestCenter = i;
                bestArmLen = armLen;
            }
        } 
        return s.substr((bestCenter - bestArmLen) / 2, bestArmLen);
    }
};

File: LeetCode/practice-rust/LeetCode-5-Manacher.rs
// 最长回文子串
struct Solution;

impl Solution {
    pub fn longest_palindrome(s: String) -> String {
        if s.is_empty() { return String::new(); }

        let mut str = vec!['#'; s.len() * 2 + 1];
        for (i, c) in s.as_bytes().iter().enumerate() {
            str[2 * i + 1] = *c as char;
        }

        let mut arm_lens = vec![0; str.len()];
        let (mut curr_center, mut curr_arm_len) = (0, 0);
        let (mut best_center, mut best_arm_len) = (0, 0);

        for (i, _) in str.iter().enumerate() {
            let mut len = 0;
            if i > curr_center + curr_arm_len {
                // nothing
            } else if i + arm_lens[2 * curr_center - i] > curr_center + curr_arm_len {
                len = curr_center + curr_arm_len - i;
            } else {
                len = arm_lens[2 * curr_center - i];
            }

            while i > len && i + len + 1 < str.len() && str[i - len - 1] == str[i + len + 1] {
                len += 1;
            }
            arm_lens[i] = len;

            if i + len > curr_center + curr_arm_len {
                curr_center = i;
                curr_arm_len = len;
            }

            if len > best_arm_len {
                best_center = i;
                best_arm_len = len;
            }
        }
        let start_index = (best_center - best_arm_len) / 2;
        s[start_index..start_index + best_arm_len].to_string()
    }
}

fn main () {
    let l1 = Solution::longest_palindrome(String::from("babad"));
    println!("{:?}", l1);
}